branch.go

```diff
@@ -99,80 +108,132 @@ func (m branchModule) createBranchHandler(c *gin.Context) {
+			defaultBranchName = reqFirstBranch.Name
 		}
 
-		var branchNames []string
-		for _, branch := range branches {
-			if branch.Default {
-				defaultBranch = branch
+		branchNamesSet := map[string]struct{}{}
+
+		for _, reqBranch := range reqBranches {
+			if reqBranch.Default {
+				defaultBranchName = reqBranch.Name
 			}
 
-			branchNames = append(branchNames, branch.Name)
-			result := m.Database.
-				Where(&branch, database.BranchFields.ProjectID, database.BranchFields.TokenID, database.BranchFields.Name).
-				First(&branch)
-			if errors.Is(result.Error, gorm.ErrRecordNotFound) {
-				if err := tx.Create(&branch).Error; err != nil {
+			branchNamesSet[reqBranch.Name] = struct{}{}
+			var count int64
+			err := m.Database.
+				Model(&database.Branch{}).
+				Where(&database.Branch{
+					ProjectID: reqBranch.ProjectID,
+					TokenID:   reqBranch.TokenID,
+					Name:      reqBranch.Name,
+				}, database.BranchFields.ProjectID, database.BranchFields.TokenID, database.BranchFields.Name).
+				Limit(1).
+				Count(&count).
+				Error
+			if err != nil {
+				return err
+			}
+			if count == 0 {
+				if err := tx.Create(&database.Branch{
+					ProjectID: reqBranch.ProjectID,
+					TokenID:   reqBranch.TokenID,
+					Name:      reqBranch.Name,
+				}).Error; err != nil {
 					return err
 				}
 			}
 		}
 
 		//set single default branch in project
 		if err := tx.
-			Model(&Branch{}).
-			Where(&Branch{ProjectID: defaultBranch.ProjectID, Default: true}).
-			Where(tx.Not(&defaultBranch, database.BranchFields.Name)).
+			Model(&database.Branch{}).
+			Where(&database.Branch{ProjectID: projectID, Default: true}).
+			Where(tx.Not(&database.Branch{Name: defaultBranchName}, database.BranchFields.Name)).
 			Select(database.BranchFields.Default).
-			Updates(&Branch{Default: false}).Error; err != nil {
+			Updates(&database.Branch{Default: false}).Error; err != nil {
 			return err
 		}
 
-		for _, oldDbBranch := range oldDbBranches {
-			if !contains(branchNames, oldDbBranch.Name) {
+		for _, dbOldBranch := range dbOldBranches {
+			if _, ok := branchNamesSet[dbOldBranch.Name]; !ok {
 				//remove old branch
 				if err := tx.
-					Where(&oldDbBranch, database.BranchFields.ProjectID, database.BranchFields.TokenID, database.BranchFields.Name).
-					Delete(&oldDbBranch).Error; err != nil {
+					Where(&dbOldBranch, database.BranchFields.ProjectID, database.BranchFields.TokenID, database.BranchFields.Name).
+					Delete(&dbOldBranch).Error; err != nil {
 					return err
 				}
 				log.Info().
-					WithString("branch", oldDbBranch.Name).
-					WithUint("project", oldDbBranch.ProjectID).
+					WithString("branch", dbOldBranch.Name).
+					WithUint("project", dbOldBranch.ProjectID).
 					Message("Deleted branch from project.")
 			}
 		}
 
-		return nil
+		return m.Database.Find(&dbNewBranches).Error
 	})
+
+	if err != nil {
+		log.Error().
+			WithError(err).
+			Message("Failed to replace branch list. Transaction reverted.")
```

Comment:
Might as well use SQL nomenclature :)

```suggestion
			Message("Failed to replace branch list. Transaction rolled back.")
```

Reply:
Very nice!